src/main/groovy/com/rundeck/plugins/ansible/plugin/AnsibleResourceModelSource.java

```diff
@@ -391,246 +404,268 @@ public INodeSet getNodes() throws ResourceModelSourceException {
-          for (JsonElement ele : root.getAsJsonArray("group_names")) {
-            if (ignoreTagPrefix != null && ignoreTagPrefix.length() > 0 && ele.getAsString().startsWith(ignoreTagPrefix)) continue;
-            tags.add(ele.getAsString());
-          }
-          // Add extraTag to node
-          if (extraTag != null && extraTag.length() > 0) {
-            tags.add(extraTag);
-          }
-          node.setTags(tags);
+    try {
+      Files.walkFileTree(tempDirectory, new SimpleFileVisitor<Path>() {
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+          Files.delete(file);
+          return FileVisitResult.CONTINUE;
+        }
 
-          if (root.has("ansible_lsb") && root.getAsJsonObject("ansible_lsb").has("description")) {
-            node.setDescription(root.getAsJsonObject("ansible_lsb").get("description").getAsString());
-          } else {
-            StringBuilder sb = new StringBuilder();
-
-            if (root.has("ansible_distribution") && !root.get("ansible_distribution").isJsonNull()) {
-              sb.append(root.get("ansible_distribution").getAsString()).append(" ");
-            }
-            if (root.has("ansible_distribution_version")) {
-              sb.append(root.get("ansible_distribution_version").getAsString()).append(" ");
-            }
-
-            if (sb.length() > 0) {
-              node.setDescription(sb.toString().trim());
-            }
-          }
+        @Override
+        public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+          Files.delete(dir);
+          return FileVisitResult.CONTINUE;
+        }
+      });
+    } catch (IOException e) {
+      throw new ResourceModelSourceException("Error deleting temporary directory.", e);
+    }
 
-          // ansible_system     = Linux   = osFamily in Rundeck
-          // ansible_os_family  = Debian  = osName in Rundeck
+    return nodes;
+  }
 
-          if (root.has("ansible_os_family")) {
-            node.setOsFamily(root.get("ansible_os_family").getAsString());
-          }
+  public void executeFutures(Stream<Path> directories, NodeSetImpl nodes) {
+    log.debug("Number of Threads: {}", numberThreads);
+    ExecutorService executor = Executors.newFixedThreadPool(Integer.parseInt(numberThreads));
 
-          if (root.has("ansible_os_name") && !root.get("ansible_os_name").isJsonNull()) {
-            node.setOsName(root.get("ansible_os_name").getAsString());
-          }
+    List<CompletableFuture<INodeEntry>> futures = directories
+            .map(factFile -> processFile(factFile, executor))
+            .collect(Collectors.toList());
 
-          if (root.has("ansible_architecture") && !root.get("ansible_architecture").isJsonNull()) {
-            node.setOsArch(root.get("ansible_architecture").getAsString());
-          }
+    futures.stream().map(CompletableFuture::join).forEach(nodes::putNode);
```

Comment:
AFAIU, with the calling thread blocked and a fixed thread pool of size 1, we would end up doing synchronous processing. 
I'm surprised to see your tests with a single thread showing perf improvements.
https://pagerduty.atlassian.net/browse/RUN-2366?focusedCommentId=747531
I might be missing something. Can you please elaborate?